snapcraft/storeapi/__init__.py

```diff
@@ -89,6 +89,22 @@ def _get_name_from_snap_file(snap_path):
     return snap_yaml['name']
 
 
+def _macaroon_auth(conf):
+    """Format a macaroon and its associated discharge.
+
+    :return: A string suitable to use in an Authorization header.
+
+    """
+    root_macaroon_raw = conf.get('macaroon')
+    unbound_raw = conf.get('unbound_discharge')
+    root_macaroon = macaroons.Macaroon.deserialize(root_macaroon_raw)
+    unbound = macaroons.Macaroon.deserialize(unbound_raw)
+    bound = root_macaroon.prepare_for_request(unbound)
+    auth = 'Macaroon root={}, discharge={}'.format(root_macaroon_raw,
+                                                   bound.serialize())
```

Comment:
I like a lot better the way that you inject the auth in the request. I hope this code gets out of snapcraft and goes into a shared tool, but if it remains here we can try to copy more things from you.

However there are a couple of details that I don't like about your implementation.
https://bazaar.launchpad.net/~launchpad-pqm/launchpad/devel/view/18086/lib/lp/snappy/model/snapstoreclient.py#L69
Here I think you shouldn't use assert. It can be disabled and you should use it only for internal consistency checks. The value seems to come from somewhere out of your control, so it should be guarded with if -> raise.
The previous check seems to be safe with the assert.

https://bazaar.launchpad.net/~launchpad-pqm/launchpad/devel/view/18086/lib/lp/snappy/tests/test_snapstoreclient.py#L98
Here you are testing only the root macaroon. The second argument, the discharge_macaroon will fail on deserialize, before it gets to the assert.

It made sense to me that any problem reading the macaroon is an InvalidCredentialsError, so I am  throwing an InvalidCredentialsError if deserialize fails with that binascii padding error.
Please let me know what do you think.